### PR TITLE
kconfig: changed cores and memory max sizes to be configurable

### PR DESCRIPTION
**Pull request by dabekjakub**
changed cores and memory max sizes to be configurable for cavs platforms

Signed-off-by: Jakub Dabek <jakub.dabek@linux.intel.com>

### CODE REVIEW

File: src/arch/xtensa/Kconfig [hunk 1]
```diff
@@ -8,4 +8,23 @@ config SMP
...
 
+config CORE_COUNT
+	int "Number of cores"
+	range 1 4
+	depends on CAVS
+	help
+	  Number of cores
+
+config HP_MEMORY_BANKS
+	int "HP memory  baks count"
```
Comment (lgirdwood) [inline src/arch/xtensa/Kconfig]: banks

File: src/arch/xtensa/Kconfig [hunk 2]
```diff
@@ -8,4 +8,23 @@ config SMP
...
+	depends on CAVS
+	help
+	  Avalible memory banks cont for HP memory
```
Comment (lgirdwood) [inline src/arch/xtensa/Kconfig]: Can we say High Performance in the help section instead of HP

File: src/arch/xtensa/Kconfig [hunk 3]
```diff
@@ -8,4 +8,23 @@ config SMP
...
+
+config LP_MEMORY_BANKS
+	int "LP memory  baks count"
+	depends on CAVS
+	help
+	  Avalible memory banks cont for LP memory
```
Comment (lgirdwood) [inline src/arch/xtensa/Kconfig]: Diito for LP, lets say Low Power

File: src/CMakeLists.txt [hunk 4]
```diff
@@ -1,3 +1,4 @@
+add_subdirectory(platform)
```
Comment (dbaluta) [inline src/CMakeLists.txt]: Please use the imperative mood in commit message. Read more here:

https://chris.beams.io/posts/git-commit/#imperative

I would also prefer to split this in two separate patches. One patch per change.

So, you will have something like this:

kconfig: Make number of cores configurable
kconfig: Make memory limits configurable

File: src/platform/apollolake/include/platform/asm_memory_management.h [hunk 5]
```diff
@@ -53,7 +53,7 @@
 	// SEGMENT #0
 	movi \az, (SHIM_BASE + SHIM_HSPGISTS)
 	movi \ax, (SHIM_BASE + SHIM_HSPGCTL)
-	movi \ay, HPSRAM_MASK()
+	movi \ay, HPSRAM_MASK(0)
 	s32i \ay, \ax, 0
```
Comment (dbaluta) [inline src/platform/apollolake/include/platform/asm_memory_management.h]: This change looks totally unrelated to the commit message. Can you make this a separate patch.

File: src/CMakeLists.txt [hunk 6]
```diff
@@ -17,5 +18,4 @@ add_subdirectory(init)
 add_subdirectory(ipc)
 add_subdirectory(lib)
 add_subdirectory(math)
-add_subdirectory(platform)
```
Comment (dbaluta) [inline src/CMakeLists.txt]: Sorry, but I don't understand this change. Why do you remove this here?

File: src/platform/Kconfig [hunk 7]
```diff
@@ -85,7 +85,20 @@ config ICELAKE
 	select CAVS
 	select CAVS_VERSION_2_0
 	help
-	  Select if your target platform is Icelake-compatible
+		Select if your target platform is Icelake-compatible
+
+config ICELAKE_MIN
```
Comment (dbaluta) [inline src/platform/Kconfig]: This definitely needs a separate patch and proper explanation.

File: src/platform/apollolake/include/platform/asm_memory_management.h (hunk 5 above)
Comment (dabekjakub) [reply src/platform/apollolake/include/platform/asm_memory_management.h]: It is related  since the  macro was moved to cavs/memory.h file and needs a parameter. This parameter is integral to the macro move.(it won't build if it is not changed)

File: src/CMakeLists.txt (hunk 4 above)
Comment (dabekjakub) [reply src/CMakeLists.txt]: Commit message - sure my bad

I cannot split it since it is interconnected and would have to do workarounds for one patch to remove theme in the other.

File: src/CMakeLists.txt (hunk 6 above)
Comment (dabekjakub) [reply src/CMakeLists.txt]: If left in this position platform is gone through as last file and the arch needs platform defines for linker/asm scripts.
Tldr: Won't build otherwise.

File: src/platform/Kconfig (hunk 7 above)
Comment (dabekjakub) [reply src/platform/Kconfig]: Agreed i will send other patch for this

File: src/CMakeLists.txt (hunk 4 above)
Comment (dbaluta) [reply src/CMakeLists.txt]: I'm pretty sure it can be done. 

1) 2 patches to introduce the Kconfigs (core/memory)
2) 2 patches (1 for core numbers / 1 for memory limits) that remove the hardcoded values and replace them with CONFIG_ introduced in patches below
3) patch that introduce ICELAKE stuff. Not sure about this because I need to see it first as a separate patch.

File: src/CMakeLists.txt (hunk 6 above)
Comment (jajanusz) [reply src/CMakeLists.txt]: This order shouldn't matter for build. Can you give some more details of what's needed by arch ?

File: src/CMakeLists.txt (hunk 4 above)
Comment (jajanusz) [reply src/CMakeLists.txt]: @dbaluta When you have comment regarding something that is not in diff (f.e. commit messages), I think it's better to have it in overall review comment instead of binding it to unrelated line.

File: src/arch/xtensa/Kconfig (hunk 2 above)
Comment (plbossart) [reply src/arch/xtensa/Kconfig]: typo: available.

File: src/arch/xtensa/Kconfig (hunk 3 above)
Comment (plbossart) [reply src/arch/xtensa/Kconfig]: available again, and it's banks not baks.

File: src/arch/xtensa/configs/cannonlake_defconfig [hunk 8]
```diff
@@ -2,3 +2,6 @@ CONFIG_SMP=y
 CONFIG_CANNONLAKE=y
 CONFIG_CAVS_DMIC=y
 CONFIG_CAVS_SSP=y
+CONFIG_CORE_COUNT=4
+CONFIG_LP_MEMORY_BANKS=1
+CONFIG_HP_MEMORY_BANKS=47
```
Comment (plbossart) [inline src/arch/xtensa/configs/cannonlake_defconfig]: can we add a boilerplate CONFIG_CAVS_SOUNDWIRE while we are at it, starting with CNL?

Comment (jajanusz) [reply src/arch/xtensa/configs/cannonlake_defconfig]: @plbossart Why? Isn't it better to introduce config when it's going to be used?

**Comment by plbossart** [comment]
> @plbossart <https://github.com/plbossart> Why? Isn't it better to 
> introduce config when it's going to be used?

'when' is next week  for me :-)

File: src/CMakeLists.txt (hunk 4 above)
Comment (dbaluta) [reply src/CMakeLists.txt]: @jajanusz sure I'm still confused sometimes about github workflow. I honestly miss review on email :)

File: src/arch/xtensa/configs/cannonlake_defconfig (hunk 8 above)
Comment (dabekjakub) [reply src/arch/xtensa/configs/cannonlake_defconfig]: @plbossart I wouldn't do SDW here since it has nothing to do with core/memory pr. I am already told that i should split it ;)

File: src/CMakeLists.txt (hunk 4 above)
Comment (dabekjakub) [reply src/CMakeLists.txt]: I may agree with intent. However You are suggesting splitting up  something that is essentially one feature. Platforms are defined by core number and available memory. One configurable without the other makes no sense to me.

File: src/CMakeLists.txt (hunk 6 above)
Comment (dabekjakub) [reply src/CMakeLists.txt]: It does since elements that are under platform sub directory are used in ASM files that won't build unless these  defines are run first.

File: src/platform/Kconfig (hunk 7 above)
Comment (dabekjakub) [reply src/platform/Kconfig]: removed in this patch

File: src/CMakeLists.txt (hunk 6 above)
Comment (jajanusz) [reply src/CMakeLists.txt]: Will be fixed in separate PR for #1235

File: src/arch/xtensa/Kconfig [hunk 9]
```diff
@@ -8,4 +8,23 @@ config SMP
 	help
 	  Build SMP instead of UP
 
+config CORE_COUNT
+	int "Number of cores"
+	range 1 4
+	depends on CAVS
```
Comment (jajanusz) [inline src/arch/xtensa/Kconfig]: can we have some defaults here? (generic default, not in defconfig)

Comment (dabekjakub) [reply src/arch/xtensa/Kconfig]: guess one core is always a must ;)

Comment (jajanusz) [reply src/arch/xtensa/Kconfig]: Making it explicit is better. Also I meant that we need default not only for this config but for all 'int' configs.

Comment (jajanusz) [reply src/arch/xtensa/Kconfig]: @mmaka1 Why does it depend on CAVS? Looks like generic config, if it's cavs-specific should it be named with CAVS_* prefix?

**Comment by dabekjakub** [comment]
@lbetlej could You take a look at this pr since some of the changes touch ASM files

**Comment by sofci** [comment]
Can one of the admins verify this patch?

File: src/arch/xtensa/Kconfig (hunk 9 above)
Comment (mmaka1) [reply src/arch/xtensa/Kconfig]: @jajanusz I agree, on this level there is no relation to CAVS platforms. This is one of the knobs the architecture provides.

File: src/arch/xtensa/Kconfig [hunk 10]
```diff
@@ -8,4 +8,26 @@ config SMP
...
+	help
+	  Number of cores
+
+config HP_MEMORY_BANKS
+	int "HP memory  banks count"
+	depends on CAVS
+	default 8
+	help
+	  Available memory banks cont for High Performace memory
```
Comment (mmaka1) [inline src/arch/xtensa/Kconfig]: count

File: src/arch/xtensa/Kconfig [hunk 11]
```diff
@@ -8,4 +8,26 @@ config SMP
...
+
+config LP_MEMORY_BANKS
+	int "LP memory  banks count"
+	default 1
+	depends on CAVS
+	help
+	  Available memory banks cont for Low Power memory
```
Comment (mmaka1) [inline src/arch/xtensa/Kconfig]: count

File: src/platform/apollolake/include/platform/platform.h [hunk 12]
```diff
@@ -33,25 +33,14 @@
 #ifndef __PLATFORM_PLATFORM_H__
 #define __PLATFORM_PLATFORM_H__
 
+#include <cavs/memory.h>
+
 #define PLATFORM_RESET_MHE_AT_BOOT		1
 
 #define PLATFORM_DISABLE_L2CACHE_AT_BOOT	1
 
-#define PLATFORM_CORE_COUNT			2
```
Comment (mmaka1) [inline src/platform/apollolake/include/platform/platform.h]: Would be good, for the sake of config verification purpose, to keep `MAX_CORE_COUNT` h/c in the platform header (as specific platform max capability) and verify against the `CORE_COUNT` defined by user in the Kconfig during compilation (throw pre-proc #error if invalid).

File: src/platform/intel/cavs/include/cavs/memory.h [hunk 13]
```diff
@@ -31,6 +31,36 @@
...
+
+#define PLATFORM_HPSRAM_SEGMENTS	((PLATFORM_HPSRAM_EBB_COUNT \
+	+ EBB_BANKS_IN_SEGMENT - 1) / EBB_BANKS_IN_SEGMENT)
+
+#define LPSRAM_MASK(ignored)	((1 << PLATFORM_LPSRAM_EBB_COUNT) - 1)
+
+#define HPSRAM_MASK(seg_idx)	((1 << (PLATFORM_HPSRAM_EBB_COUNT \
+	- EBB_BANKS_IN_SEGMENT * seg_idx)) - 1)
+
+#define LPSRAM_BANK_SIZE (64 * 1024)
```
Comment (mmaka1) [inline src/platform/intel/cavs/include/cavs/memory.h]: #define LPSRAM_BANK_SIZE as SRAM_BANK_SIZE

File: src/arch/xtensa/Kconfig [hunk 14]
```diff
@@ -8,4 +8,26 @@ config SMP
...
 
+config CORE_COUNT
+	int "Number of cores"
+	range 1 4
+	default 1
+	depends on CAVS
+	help
+	  Number of cores
+
+config HP_MEMORY_BANKS
```
Comment (mmaka1) [inline src/arch/xtensa/Kconfig]: While the number of cores is cpu arch related, the L2 memory is platform specific, interfaced with cpu through PIF. So move the memory defined into the CAVS submenu.

**Comment by dabekjakub** [comment]
@lgirdwood all changes implemented

File: Kconfig [hunk 15]
```diff
@@ -2,6 +2,24 @@ mainmenu "SOF $(PROJECTVERSION) Configuration"
...
 comment "Compiler: $(CC_VERSION_TEXT)"
 
+menu "CAVS"
+
+config HP_MEMORY_BANKS
+	int "HP memory banks count"
+	depends on CAVS
+	default 8
+	help
+	  Available memory banks count for High Performace memory
```
Comment (lgirdwood) [inline Kconfig]: One minor change, I would add here more detail on the memory banks. i.e. the bank size and that disabling unused banks would lower power consumption. Ditto for cores.

Comment (lgirdwood) [reply Kconfig]: also worth mentioning that reducing the number of banks too low could result in not enough free space for base FW image.

Comment (dabekjakub) [reply Kconfig]: Added

File: src/platform/Kconfig [hunk 16]
```diff
@@ -85,7 +85,20 @@ config ICELAKE
 	select CAVS
 	select CAVS_VERSION_2_0
 	help
-	  Select if your target platform is Icelake-compatible
+		Select if your target platform is Icelake-compatible
```
[the remark this answers is not in the document]
Comment (plbossart) [reply src/platform/Kconfig]: indent looks indeed wrong. it's typically 2 spaces after the help.

File: Kconfig [hunk 17]
```diff
@@ -2,6 +2,27 @@ mainmenu "SOF $(PROJECTVERSION) Configuration"
...
```
Comment (tlauda) [inline Kconfig]: Performance

File: Kconfig [hunk 18]
```diff
@@ -2,6 +2,27 @@ mainmenu "SOF $(PROJECTVERSION) Configuration"
...
+	  Lowering avalible banks could result in lower power consumption
```
Comment (tlauda) [inline Kconfig]: available

File: Kconfig [hunk 19]
```diff
@@ -2,6 +2,27 @@ mainmenu "SOF $(PROJECTVERSION) Configuration"
...
+	  Too low count sould result in unresponsive/crashing image due to not
```
Comment (tlauda) [inline Kconfig]: should

File: src/arch/xtensa/Kconfig [hunk 20]
```diff
@@ -8,4 +8,10 @@ config SMP
 	help
 	  Build SMP instead of UP
 
+config CORE_COUNT
+	int "Number of cores"
+	default 1
+	help
+	  Number of used cores
+	  Lowering avalible core count could result in lower power consumption
```
Comment (tlauda) [inline src/arch/xtensa/Kconfig]: available

File: Kconfig (hunk 18 above)
Comment (lgirdwood) [reply Kconfig]: @dabekjakub it's also missing the block sizes IIRC 64k, this makes the number of blocks option more meaningful.
